### PR TITLE
save HLT objects associated with HLT_Ele27_WPTight_Gsf_v and pt>25

### PR DESCRIPTION
**Pull request by vhegde91**

### CODE REVIEW

File: Utils/src/TriggerProducer.cc [hunk 1]
```diff
@@ -48,6 +50,9 @@ class TriggerProducer : public edm::global::EDProducer<> {
   edm::EDGetTokenT<edm::TriggerResults> trigResultsTok_;
   edm::EDGetTokenT<pat::PackedTriggerPrescales> trigPrescalesTok_;
   std::vector<std::string> parsedTrigNamesVec;
+  edm::EDGetTokenT<edm::TriggerResults> trigResultsToken;
```
Comment (kpedro88) [inline Utils/src/TriggerProducer.cc]: we already have `trigResultsTok_`, don't need another one

File: Utils/src/TriggerProducer.cc [hunk 2]
```diff
@@ -89,9 +94,15 @@ TriggerProducer::TriggerProducer(const edm::ParameterSet& iConfig)
   trigResultsTok_ = consumes<edm::TriggerResults>(trigResultsTag_);
   trigPrescalesTok_ = consumes<pat::PackedTriggerPrescales>(trigPrescalesTag_);
 
+  edm::InputTag theTriggerLabel("TriggerResults", "", "HLT");
```
Comment (kpedro88) [inline Utils/src/TriggerProducer.cc]: remove

File: Utils/src/TriggerProducer.cc [hunk 3]
```diff
@@ -89,9 +94,15 @@ TriggerProducer::TriggerProducer(const edm::ParameterSet& iConfig)
...
+  edm::InputTag theTrigObjLabel("selectedPatTrigger");
```
Comment (kpedro88) [inline Utils/src/TriggerProducer.cc]: make configurable

File: Utils/src/TriggerProducer.cc [hunk 4]
```diff
@@ -89,9 +94,15 @@ TriggerProducer::TriggerProducer(const edm::ParameterSet& iConfig)
...
+  trigResultsToken = consumes<edm::TriggerResults>(theTriggerLabel);
```
Comment (kpedro88) [inline Utils/src/TriggerProducer.cc]: remove

File: Utils/src/TriggerProducer.cc [hunk 5]
```diff
@@ -89,9 +94,15 @@ TriggerProducer::TriggerProducer(const edm::ParameterSet& iConfig)
...
+  trigObjCollToken = consumes<pat::TriggerObjectStandAloneCollection>(theTrigObjLabel);
+
   produces<std::vector<std::string> >("TriggerNames");
   produces<std::vector<int> >("TriggerPass");
   produces<std::vector<int> >("TriggerPrescales");
+  produces<std::vector<TLorentzVector> >("HLTElectronObjects");
```
Comment (kpedro88) [inline Utils/src/TriggerProducer.cc]: add a configurable bool so it doesn't produce this all the time, only when requested

File: Utils/src/TriggerProducer.cc [hunk 6]
```diff
@@ -150,9 +162,36 @@ TriggerProducer::produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetu
     }
   }
 
+  edm::Handle<edm::TriggerResults> triggerBits;
+  iEvent.getByToken(trigResultsToken,triggerBits);
```
Comment (kpedro88) [inline Utils/src/TriggerProducer.cc]: remove these two lines

File: Utils/src/TriggerProducer.cc [hunk 7]
```diff
@@ -150,9 +162,36 @@ TriggerProducer::produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetu
...
+
+  const edm::TriggerNames& names = iEvent.triggerNames(*triggerBits);
```
Comment (kpedro88) [inline Utils/src/TriggerProducer.cc]: remove this, use `trigNames` defined above

File: Utils/src/TriggerProducer.cc [hunk 8]
```diff
@@ -150,9 +162,36 @@ TriggerProducer::produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetu
...
+  edm::Handle<pat::TriggerObjectStandAloneCollection> triggerObjects;
```
Comment (kpedro88) [inline Utils/src/TriggerProducer.cc]: put all of this behind bool

File: Utils/src/TriggerProducer.cc [hunk 9]
```diff
@@ -150,9 +162,36 @@ TriggerProducer::produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetu
...
+  iEvent.getByToken(trigObjCollToken,triggerObjects);
+  //save the trigger object corresponding to the trigger HLT_Ele27_WPTight_Gsf_v*. Obtained code from https://twiki.cern.ch/twiki/bin/view/CMSPublic/WorkBookMiniAOD2016#Trigger
+  // loop over selected trigger objects                                                                                              
+  for (pat::TriggerObjectStandAlone obj : *triggerObjects) {
+    obj.unpackPathNames(names);
+    if(obj.pt() < 25.0) continue;//look for HLT objects with Pt > 25GeV only.
```
Comment (kpedro88) [inline Utils/src/TriggerProducer.cc]: do this before unpacking

File: Utils/src/TriggerProducer.cc [hunk 10]
```diff
@@ -150,9 +162,36 @@ TriggerProducer::produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetu
...
+    TLorentzVector objVec(obj.px(),obj.py(),obj.pz(),obj.energy());
```
Comment (kpedro88) [inline Utils/src/TriggerProducer.cc]: remove this line

File: Utils/src/TriggerProducer.cc [hunk 11]
```diff
@@ -150,9 +162,36 @@ TriggerProducer::produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetu
...
+    std::vector<std::string> pathNamesAll = obj.pathNames(false);
+    std::vector<std::string> pathNamesLast = obj.pathNames(true);
+    for (unsigned h = 0, n = pathNamesAll.size(); h < n; ++h) {
+      bool isBoth = obj.hasPathName( pathNamesAll[h], true, true );//object is associated wih l3 filter and associated to the last filter of a successfull path. this object caused the trigger to fire.
+      std::string path_i = pathNamesAll[h];
+      path_i.pop_back(); path_i.pop_back(); //remove last 2 characters from path name.
+      if( isBoth && ( (path_i == "HLT_Ele27_WPTight_Gsf_v") || (path_i == "HLT_Ele27_WPTight_Gsf_") ) ){ //check if the trigger path name is "HLT_Ele27_WPTight_Gsf_vX" or "HLT_Ele27_WPTight_Gsf_vXX"
+	hltEleObj->push_back(objVec);//save the trigger object corresponding to HLT_Ele27_WPTight_Gsf_v* trigger.
```
Comment (kpedro88) [inline Utils/src/TriggerProducer.cc]: use `emplace_back(obj.px(),obj.py(),obj.pz(),obj.energy())`

File: Utils/src/TriggerProducer.cc [hunk 12]
```diff
@@ -150,9 +162,36 @@ TriggerProducer::produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetu
...
+  iEvent.getByToken(trigObjCollToken,triggerObjects);
+  //save the trigger object corresponding to the trigger HLT_Ele27_WPTight_Gsf_v*. Obtained code from https://twiki.cern.ch/twiki/bin/view/CMSPublic/WorkBookMiniAOD2016#Trigger
+  // loop over selected trigger objects                                                                                              
+  for (pat::TriggerObjectStandAlone obj : *triggerObjects) {
+    obj.unpackPathNames(names);
+    if(obj.pt() < 25.0) continue;//look for HLT objects with Pt > 25GeV only.
+    TLorentzVector objVec(obj.px(),obj.py(),obj.pz(),obj.energy());
+    std::vector<std::string> pathNamesAll = obj.pathNames(false);
+    std::vector<std::string> pathNamesLast = obj.pathNames(true);
+    for (unsigned h = 0, n = pathNamesAll.size(); h < n; ++h) {
```
Comment (kpedro88) [inline Utils/src/TriggerProducer.cc]: `(unsigned h = 0; h < pathNamesAll.size(); ++h)`

File: Utils/src/TriggerProducer.cc [hunk 13]
```diff
@@ -150,9 +162,36 @@ TriggerProducer::produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetu
...
+  const edm::TriggerNames& names = iEvent.triggerNames(*triggerBits);
+  edm::Handle<pat::TriggerObjectStandAloneCollection> triggerObjects;
+  iEvent.getByToken(trigObjCollToken,triggerObjects);
+  //save the trigger object corresponding to the trigger HLT_Ele27_WPTight_Gsf_v*. Obtained code from https://twiki.cern.ch/twiki/bin/view/CMSPublic/WorkBookMiniAOD2016#Trigger
+  // loop over selected trigger objects                                                                                              
+  for (pat::TriggerObjectStandAlone obj : *triggerObjects) {
+    obj.unpackPathNames(names);
+    if(obj.pt() < 25.0) continue;//look for HLT objects with Pt > 25GeV only.
+    TLorentzVector objVec(obj.px(),obj.py(),obj.pz(),obj.energy());
+    std::vector<std::string> pathNamesAll = obj.pathNames(false);
```
Comment (kpedro88) [inline Utils/src/TriggerProducer.cc]: `const std::vector<std::string>&`

File: Utils/src/TriggerProducer.cc [hunk 14]
```diff
@@ -150,9 +162,36 @@ TriggerProducer::produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetu
...
+    std::vector<std::string> pathNamesLast = obj.pathNames(true);
```
Comment (kpedro88) [inline Utils/src/TriggerProducer.cc]: `const std::vector<std::string>&`

File: Utils/src/TriggerProducer.cc [hunk 15]
```diff
@@ -150,9 +162,36 @@ TriggerProducer::produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetu
...
+    for (unsigned h = 0, n = pathNamesAll.size(); h < n; ++h) {
+      bool isBoth = obj.hasPathName( pathNamesAll[h], true, true );//object is associated wih l3 filter and associated to the last filter of a successfull path. this object caused the trigger to fire.
+      std::string path_i = pathNamesAll[h];
+      path_i.pop_back(); path_i.pop_back(); //remove last 2 characters from path name.
```
Comment (kpedro88) [inline Utils/src/TriggerProducer.cc]: remove this and in the condition on the next line, just use: `path_i.find("HLT_Ele27_WPTight_Gsf_v")!= std::string::npos`

File: Utils/src/TriggerProducer.cc [hunk 16]
```diff
@@ -150,9 +162,36 @@ TriggerProducer::produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetu
...
+  // loop over selected trigger objects                                                                                              
+  for (pat::TriggerObjectStandAlone obj : *triggerObjects) {
+    obj.unpackPathNames(names);
+    if(obj.pt() < 25.0) continue;//look for HLT objects with Pt > 25GeV only.
+    TLorentzVector objVec(obj.px(),obj.py(),obj.pz(),obj.energy());
+    std::vector<std::string> pathNamesAll = obj.pathNames(false);
+    std::vector<std::string> pathNamesLast = obj.pathNames(true);
+    for (unsigned h = 0, n = pathNamesAll.size(); h < n; ++h) {
+      bool isBoth = obj.hasPathName( pathNamesAll[h], true, true );//object is associated wih l3 filter and associated to the last filter of a successfull path. this object caused the trigger to fire.
+      std::string path_i = pathNamesAll[h];
```
Comment (kpedro88) [inline Utils/src/TriggerProducer.cc]: `const std::string&`

File: Utils/src/TriggerProducer.cc [hunk 17]
```diff
@@ -150,9 +162,36 @@ TriggerProducer::produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetu
...
+	hltEleObj->push_back(objVec);//save the trigger object corresponding to HLT_Ele27_WPTight_Gsf_v* trigger.
+	break;
+      }
+    }
+  }
+
   iEvent.put(std::move(passTrigVec),"TriggerPass");
   iEvent.put(std::move(trigPrescaleVec),"TriggerPrescales");
   iEvent.put(std::move(trigNamesVec),"TriggerNames");
+  iEvent.put(std::move(hltEleObj),"HLTElectronObjects");
```
Comment (kpedro88) [inline Utils/src/TriggerProducer.cc]: put behind bool

File: TreeMaker/python/makeTreeFromMiniAOD_cff.py [hunk 18]
```diff
@@ -495,6 +495,7 @@ def makeTreeFromMiniAOD(self,process):
     )
     self.VectorInt.extend(['TriggerProducer:TriggerPass','TriggerProducer:TriggerPrescales'])
     self.VectorString.extend(['TriggerProducer:TriggerNames'])
+    self.VectorTLorentzVector.extend(['TriggerProducer:HLTElectronObjects'])
```
Comment (kpedro88) [inline TreeMaker/python/makeTreeFromMiniAOD_cff.py]: condition on `if "SingleElectron" in process.source.fileNames[0]:`

**Review by kpedro88 — CHANGES_REQUESTED** [review]

File: Utils/src/TriggerProducer.cc (hunk 12 above)
Comment (kpedro88) [reply Utils/src/TriggerProducer.cc]: actually, just do: `(const auto& pathName : pathNamesAll)`

File: Utils/src/TriggerProducer.cc (hunk 14 above)
Comment (kpedro88) [reply Utils/src/TriggerProducer.cc]: actually, seems like this variable isn't used, so just delete it

File: Utils/src/TriggerProducer.cc (hunk 5 above)
Comment (vhegde91) [reply Utils/src/TriggerProducer.cc]: did you mean something like this?
bool saveHLTObj = true or false;
if(saveHLTObj) produces<std::vector<TLorentzVector> >("HLTElectronObjects");

this bool will be used in the next part of the code as well.

Comment (kpedro88) [reply Utils/src/TriggerProducer.cc]: yes, and it should be set by the python config for the module.

**Comment by vhegde91** [comment]
updated the code with most of your suggestions.
sorry, tried to make it make configurable from python, but could not

**Comment by kpedro88** [comment]
@vhegde91 thanks
it should be easy to make it configurable: just do
```
saveHLTObj = iConfig.getParameter<bool>("saveHLTObj");
```
then in https://github.com/TreeMaker/TreeMaker/blob/Run2/Utils/python/triggerproducer_cfi.py, add:
```
saveHLTObj = cms.bool(False),
```
and in `makeTreeFromMiniAOD`, in `process.TriggerProducer`:
```
    saveHLTObj = cms.bool("SingleElectron" in process.source.fileNames[0])
```

File: Utils/src/TriggerProducer.cc [hunk 19]
```diff
@@ -150,9 +159,29 @@ TriggerProducer::produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetu
     }
   }
 
+  edm::Handle<pat::TriggerObjectStandAloneCollection> triggerObjects;
```
Comment (kpedro88) [inline Utils/src/TriggerProducer.cc]: put this entire section inside `if(saveHLTObj)`
(from here till end of for loop)

**Comment by vhegde91** [comment]
@kpedro88 thanks a lot for helping with this. let me know if anything else is left.

**Review by kpedro88 — APPROVED** [review]

**Pull request merged**